lib/SILOptimizer/Utils/PartialApplyCombiner.cpp

```diff
@@ -35,192 +32,105 @@ class PartialApplyCombiner {
-    //
-    // TODO: This is because we are inserting alloc_stack at the beginning/end
-    // of functions where the dependent type may not exist yet.
-    if (arg->getType().hasOpenedExistential())
-      return false;
-
-    // If the temporary is non-trivial, we need to destroy it later.
-    if (!arg->getType().isTrivial(*pai->getFunction()))
-      needsDestroys = true;
-    argsToHandle.push_back(std::make_pair(arg, i));
+  // Of course we must inlude all apply instructions which we want to optimize.
+  for (FullApplySite ai : applies) {
+    paiUsers.push_back(ai.getInstruction());
   }
 
-  if (needsDestroys) {
-    // Compute the set of endpoints, which will be used to insert destroys of
-    // temporaries. This may fail if the frontier is located on a critical edge
-    // which we may not split (no CFG changes in SILCombine).
-    ValueLifetimeAnalysis vla(pai);
-    if (!vla.computeFrontier(partialApplyFrontier,
-                             ValueLifetimeAnalysis::DontModifyCFG))
-      return false;
-  }
+  // Also include all destroys in the liferange for the arguments.
+  // This is needed for later processing in tryDeleteDeadClosure: in case the
+  // pai gets dead after this optimization, tryDeleteDeadClosure relies on that
+  // we already copied the pai arguments to extend their lifetimes until the pai
+  // is finally destroyed.
+  collectDestroys(pai, paiUsers);
 
-  for (auto argWithIdx : argsToHandle) {
-    SILValue Arg = argWithIdx.first;
-    builder.setInsertionPoint(pai->getFunction()->begin()->begin());
-    // Create a new temporary at the beginning of a function.
-    SILDebugVariable dbgVar(/*Constant*/ true, argWithIdx.second);
-    auto *tmp = builder.createAllocStack(pai->getLoc(), Arg->getType(), dbgVar);
-    builder.setInsertionPoint(pai);
-    // Copy argument into this temporary.
-    builder.createCopyAddr(pai->getLoc(), Arg, tmp, IsTake_t::IsNotTake,
-                           IsInitialization_t::IsInitialization);
+  ValueLifetimeAnalysis vla(pai, paiUsers);
+  ValueLifetimeAnalysis::Frontier partialApplyFrontier;
 
-    tmpCopies.push_back(tmp);
-    argToTmpCopy.insert(std::make_pair(Arg, tmp));
+  // Computing the frontier may fail if the frontier is located on a critical
+  // edge which we may not split.
+  if (!vla.computeFrontier(partialApplyFrontier,
+                           ValueLifetimeAnalysis::DontModifyCFG)) {
+    return false;
   }
-  return true;
-}
 
-/// Emit dealloc_stack for all temporaries.
-void PartialApplyCombiner::deallocateTemporaries() {
-  // Insert dealloc_stack instructions at all function exit points.
-  for (SILBasicBlock &block : *pai->getFunction()) {
-    TermInst *term = block.getTerminator();
-    if (!term->isFunctionExiting())
-      continue;
+  for (Operand *argOp : argsToHandle) {
+    SILValue arg = argOp->get();
+    int argIdx = argOp->getOperandNumber() - pai->getArgumentOperandNumber();
```

Comment:
Please don't hard code this. This should be sort sort of higher level API on function conventions or ApplySite.